state/src/skip_list.rs

```diff
@@ -0,0 +1,771 @@
+
+pub fn insert_item<T, D, R>(
+    account: AccountAddress,
+    element: &T,
+    trie: &mut EthTrie<D>,
+    rng: &mut R,
+) -> Result<(), TrieError>
+where
+    T: Clone + Listable + Ord + Serialize + DeserializeOwned + 'static,
+    D: DB,
+    R: Rng,
+{
+    let insert_key = SkipListKey::new(account, 0, element);
+
+    // The key is already present, so there is nothing to do.
+    if insert_key.trie_value(trie)?.is_some() {
+        return Ok(());
+    }
+
+    let head_key = SkipListHeadKey::<T>::new(account);
+    let trie_head_key = head_key.key_hash();
+    let head = SkipListHeadValue::<T>::read_trie(&trie_head_key, trie)?;
+    let end_of_list = SkipListValue::<T> { next_value: None }.serialize();
+
+    let Some(first_value) = head.first_value else {
+        // The list is empty, so we just insert this value
+        let new_head = SkipListHeadValue::new(0, element);
+        trie.insert(trie_head_key.as_slice(), &new_head.serialize())?;
+
+        trie.insert(insert_key.key_hash().as_slice(), &end_of_list)?;
+
+        return Ok(());
+    };
+
+    if element < first_value.as_ref() {
+        // The new element needs to become the head of the list.
+
+        // Randomly pick the height for the old head (now second element).
+        let insert_level = (pick_insert_height(rng) - 1) as u32;
+
+        // Update head value
+        let max_levels = std::cmp::max(insert_level as u32, head.max_levels);
+        let new_head = SkipListHeadValue::new(max_levels, element);
+        trie.insert(trie_head_key.as_slice(), &new_head.serialize())?;
+
+        // Insert new keys for new head element
+        let insert_value = SkipListValue::<T> {
+            next_value: Some(Cow::Borrowed(&first_value)),
+        }
+        .serialize();
+        for level in 0..=insert_level {
+            let fv_key = SkipListKey::new(account, level, first_value.as_ref());
+            let insert_key = SkipListKey::new(account, level, element);
+
+            let fv_value = fv_key.trie_value(trie)?;
+            match fv_value {
+                Some(_) => trie.insert(insert_key.key_hash().as_slice(), &insert_value)?,
+                None => {
+                    trie.insert(insert_key.key_hash().as_slice(), &insert_value)?;
+                    trie.insert(fv_key.key_hash().as_slice(), &end_of_list)?;
+                }
+            }
+        }
+        for level in (insert_level + 1)..=max_levels {
+            let fv_key = SkipListKey::new(account, level, first_value.as_ref());
+            let insert_key = SkipListKey::new(account, level, element);
+
+            let fv_value = fv_key.trie_value(trie)?;
+            match fv_value {
+                Some(value) => {
+                    trie.remove(fv_key.key_hash().as_slice())?;
+                    trie.insert(insert_key.key_hash().as_slice(), &value.serialize())?;
+                }
+                None => trie.insert(insert_key.key_hash().as_slice(), &end_of_list)?,
+            }
+        }
```

Comment:
So we completely replace the head node on levels higher than its picked (re)insertion level, and prepend it with the new node on all lower ones, right? I believe this snippet would benefit from having common parts taken out of the matches, like insertion of the new first node and old first key/value construction.

Reply:
I tried to clarify things in https://github.com/UmiNetwork/op-move/pull/481/commits/4d9bd0bd0844ce610a8bba77b35f3c3f298c69fc